src/transforms/filter.jl

```diff
@@ -0,0 +1,44 @@
+# ------------------------------------------------------------------
+# Licensed under the MIT License. See LICENSE in the project root.
+# ------------------------------------------------------------------
+
+"""
+    Filter(f)
+
+The transform that filters the rows based on a given function.
+"""
+struct Filter{F} <: Transform
+  pred::F
+end
+
+isrevertible(::Type{Filter}) = true
+
+function apply(transform::Filter, input_table)
+  # Converting to Tables.rowtable to allow length function and indexing 
+  table = Tables.rows(input_table)
+
+  l = length(table)
+  f = transform.pred
+
+  # Get indices of the desired rows
+  indices = [i for i in range(1,length=l) if f(table[i])]
+
+  # Return:
+  # - the desired rows as the new table,
+  # - indices of these rows and the remaining rows as cache
+  new_table = table[indices]
+  new_table |> Tables.materializer(input_table)
+  rem_rows = table[setdiff(1:length(table), indices)]
+  new_table, (indices, rem_rows)
+end
+
+function revert(::Type{Filter}, newtable, cache)
+  indices = copy(cache[1])
+  orgtable = copy(cache[2])
+
+  for i in range(1, length=length(indices))
+    insert!(orgtable, indices[i], newtable[i])
+  end
+
+  orgtable
```

Comment:
This code is unnecessarily complicated. We rarely need explicit calls to `copy` in Julia. Can youy please refactor and return a new table that is filled row by row with the saved indices?

Reply:
`copy` was necessary, otherwise cache variable was changing every time I called the revert function. As a result, every revert function creates even bigger table, which obviously is not the goal. I can delete the first two lines and rewrite the function as follows:
```{julia}
function revert(::Type{Filter}, newtable, cache)
  for i in range(1, length=length(cache[1]))
    insert!(cache[2], cache[1][i], newtable[i])
  end

  cache[2]
end
```